src/Swarm/Game/Step/Pathfinding.hs

```diff
@@ -0,0 +1,137 @@
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Implementation of the @path@ command for robots.
+--
+-- = Design considerations
+-- One possible design of the @path@ command entailed storing a computed
+-- shortest path and providing a mechanism to retrieve parts of it later
+-- without recomputing the whole thing.
+-- However, in general the playfield can be dynamic and obstructions may
+-- appear that invalidate a given computed shortest path.
+-- Therefore, there can be limited value in caching a computed path for use
+-- across ticks.
+--
+-- Instead, in the current implementation a complete path is computed
+-- internally upon invoking the @path@ command, and just the direction of the
+-- first "move" along that path is returned as a result to the caller.
+--
+-- == Max distance
+--
+-- We allow the caller to supply a max distance, but also impose an internal maximum
+-- distance to prevent programming errors from irrecoverably freezing the game.
+module Swarm.Game.Step.Pathfinding where
+
+import Control.Carrier.State.Lazy
+import Control.Effect.Lens
+import Control.Monad (filterM, guard)
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.Maybe (MaybeT (..), runMaybeT)
+import Data.Graph.AStar (aStarM)
+import Data.HashSet (HashSet)
+import Data.HashSet qualified as HashSet
+import Data.Int (Int32)
+import Swarm.Game.Entity
+import Swarm.Game.Location
+import Swarm.Game.State
+import Swarm.Game.Step.Util
+import Swarm.Game.Universe
+import Swarm.Language.Syntax
+import Swarm.Util (hoistMaybe)
+
+-- | Shortest paths can either be computed to the nearest entity of
+-- a given type or to a specific location.
+data PathfindingTarget
+  = LocationTarget Location
+  | -- | Note: navigation to entities does not benefit from the
+    -- distance heuristic optimization of the A* algorithm.
+    EntityTarget EntityName
+
+-- | swarm command arguments are converted to idiomatic Haskell
+-- types before invoking this function, and conversely the callsite
+-- is also responsible for translating the output type to a swarm value.
+--
+-- The cost function is uniformly @1@ between adjacent cells.
+--
+-- Viable paths are determined by walkability.
+-- If the goal type is an Entity, than it is permissible for that
+-- entity to be 'Unwalkable'.
+pathCommand ::
+  (HasRobotStepState sig m, Has (State GameState) sig m) =>
+  -- | Distance limit
+  Maybe Int32 ->
```

Comment:
AFAIK The swarm numbers are unbounded Integers, so I guess this is Location implementation detail leaking into other code.

Could you instead turn manhattan distance toInteger?